### PR TITLE
Added directory for CheckPoint Store using Redis and pom.xml in it

### PR DESCRIPTION
**Pull request by anushkasingh16**
# Description

- Added a new directory with the path "..\sdk\eventhubs\azure-messaging-eventhubs-checkpointstore-redis"
- Added pom.xml file to this directory 
- Added some dependencies to pom.xml (excluding the test dependencies)"
- Added redis.clients dependency in "..\eng\versioning\external_dependencies.txt"

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by alzimmermsft — CHANGES_REQUESTED** [review]
There are a few additional files you'll need to change to hook this up:

- [ ] Add the new library into /eng/versioning/version_client.txt
  - When a library is in its first beta the dependency and current versions will be the same.
- [x] In the service directory root (/sdk/eventhubs) POM this project will need to be added as a module in the default profile and as a dependency in the code coverage profile.
- [x] A version tag is missing for the parent POM reference in the new POM file.
- [x] The POM file isn't valid as it's missing a closing XML tag.
- [x] The new project will need to be added into ci.yml and tests.yml in the EventHubs service directory
  - YML files are like Python where indentation spaces matter.

File: sdk/eventhubs/azure-messaging-eventhubs/README.md [hunk 1]
```diff
@@ -342,7 +342,6 @@ EventProcessorClient eventProcessorClient = new EventProcessorClientBuilder()
     .processEvent(eventContext -> {
         System.out.println("Partition id = " + eventContext.getPartitionContext().getPartitionId() + " and "
             + "sequence number of event = " + eventContext.getEventData().getSequenceNumber());
-        eventContext.updateCheckpoint(); // Update checkpoint for this partition.
```
Comment (liukun-msft) [inline sdk/eventhubs/azure-messaging-eventhubs/README.md:345]: Can I know why we remove `updateCheckpoint()`?

Comment (anushkasingh16) [reply sdk/eventhubs/azure-messaging-eventhubs/README.md:345]: That was an error from my side. It should be fixed now.

Comment (anushkasingh16) [reply sdk/eventhubs/azure-messaging-eventhubs/README.md:345]: That was an error from my side. Sorry about it, it has been fixed now.

File: sdk/eventhubs/azure-messaging-eventhubs/README.md [hunk 2]
```diff
@@ -342,8 +342,8 @@ EventProcessorClient eventProcessorClient = new EventProcessorClientBuilder()
...
     })
+    eventContext.updateCheckpoint(); // Update checkpoint for this partition.
```
Comment (liukun-msft) [inline sdk/eventhubs/azure-messaging-eventhubs/README.md:346]: move before `})`

File: sdk/eventhubs/azure-messaging-eventhubs/README.md (hunk 1 above)
Comment (liukun-msft) [reply sdk/eventhubs/azure-messaging-eventhubs/README.md:345]: got it.

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package com.azure.messaging.eventhubs.checkpointstore.redis;
```
Comment (alzimmermsft) [inline sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java:1]: A couple minor things here:

- All files we author must contain a Microsoft licensing header, generally it is the following:

```
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
```

I'm not sure which IDE you're using, but if it is IntelliJ there are configurations for `File and Code Templates` where you can configure all new files to automatically include this header.

- Package name is indicative of the file structure, so in this case I'd have expected this file to be contained in `/src/com/azure/messaging/eventhubs/checkpointstore/redis`. I'd recommend creating that as it'll be the base point for all sub-packages such as `com.azure.messaging.eventhubs.checkpointstore.redis`. If you're using IntelliJ there should be a right click option to add a package to a source folder, I always do each individually instead of one `com.azure.messaging.eventhubs.checkpointstore.redis` as it sometimes will create a single folder with that name which is different than subfolders but shows up the same in the UI.

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+/*
+Added this file so that an execuatable jar is produced for the build.
+ */
+
+/**
+ * Implementation of CheckpointStore that uses Azure Redis Cache.
+ */
+public class RedisCheckpointStore implements  CheckpointStore{
+    public Flux<PartitionOwnership> claimOwnership(List<PartitionOwnership> requestedPartitionOwnerships){
```
Comment (alzimmermsft) [inline sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java:17]: All public and protected classes, interfaces, methods, and fields that aren't in an `implementation` package should have javadocs,  and generally it's best to include javadocs even if they're in `implementation`.

**Review by alzimmermsft — COMMENTED** [review]
When you create the required folder structure you'll also need to add a `package-info.java` file to it.

**Review by conniey — APPROVED** [review]
LGTM. Thanks!

File: sdk/eventhubs/azure-messaging-eventhubs/README.md (hunk 2 above)
Comment (anushkasingh16) [reply sdk/eventhubs/azure-messaging-eventhubs/README.md]: Rolled it back, so hopefully it's fixed now. Sorry about that!

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java (hunk 3 above)
Comment (anushkasingh16) [reply sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java]: Thank you, this helped!

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java (hunk 4 above)
Comment (anushkasingh16) [reply sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/Main.java]: Thanks! I'll keep it in mind for next time too.

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/main/java/com/azure/messaging/eventhubs/checkpointstore/redis/JedisRedisCheckpointStore.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import com.azure.messaging.eventhubs.models.Checkpoint;
+import com.azure.messaging.eventhubs.models.PartitionOwnership;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+
+/*
+Added this file so that an execuatable jar is produced for the build.
+ */
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/main/java/com/azure/messaging/eventhubs/checkpointstore/redis/JedisRedisCheckpointStore.java:15]: nit: remove this comment

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json [hunk 6]
```diff
@@ -255,6 +255,12 @@
           "new": "class com.azure.messaging.eventhubs.EventData",
           "superClass": "com.azure.core.models.MessageContent",
           "justification": "To support Schema Registry Avro Serializer's signature, EventData extends from MessageContent."
+        },
+        {
+          "regex": true,
+          "code": "java.class.externalClassExposedInAPI",
+          "new": "class com\\.azure\\.messaging\\.eventhubs\\..*",
+          "justification": "SDK classes are allowed to be exposed by dependencies using them."
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/resources/revapi/revapi.json:263]: Why is this required?

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/main/java/module-info.java [hunk 7]
```diff
@@ -0,0 +1,5 @@
+module com.azure.messaging.eventhubs.checkpointstore.redis {
+    requires transitive com.azure.core;
+    requires redis.clients.jedis;
+    requires com.azure.messaging.eventhubs;
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/main/java/module-info.java:4]: We should also `export` all public packages in this library.

`exports com.azure.messaging.eventhubs.checkpointstore.redis`

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/pom.xml [hunk 8]
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright (c) Microsoft Corporation. All rights reserved.
+  ~ Licensed under the MIT License.
+  -->
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/pom.xml:5]: nit: this should be the first line of the file.

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json (hunk 6 above)
Comment (anushkasingh16) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json:263]: Without adding this in, the project was not being built. The error that was shared was "A class from supplementary archives is used in a public capacity in the API", to address that we exposed the APIs from Jedis. Is there any alternative way to approach to fix that error?

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/pom.xml (hunk 8 above)
Comment (anushkasingh16) [reply sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/pom.xml:5]: I locally ran into errors when I put the xml tag after this, saying that "Xml declaration should precede all document content" which is why I placed it this way. Is there a way to address this?

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/main/java/module-info.java (hunk 7 above)
Comment (anushkasingh16) [reply sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-redis/src/main/java/module-info.java:7]: Should I be making this export statement from the file BlobCheckpointStore.java? The implementation is still under construction which is why I was not sure it should be exported or not. Any advice is appreciated!

File: eng/code-quality-reports/src/main/resources/revapi/revapi.json (hunk 6 above)
Comment (alzimmermsft) [reply eng/code-quality-reports/src/main/resources/revapi/revapi.json:263]: @srnagar let's file a follow up issue to further discuss when exposing APIs from other libraries should be suppressed. Right now, we suppress Core APIs, Jackson, Reactor, and a few other external dependencies, and this looks like we need to be broader when it comes to exposing APIs from other Azure SDKs.

**Review by alzimmermsft — APPROVED** [review]
